<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2015 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Part 1: Data Overview

In Part 1 of the codelab, we perform some queries to acquaint ourselves with the data and determine whether it has any characteristics requiring any additional consideration in the QC checks that follow.

* [Variants](#variants)
* [Non-Variant Segments](#non-variant-segments)
* [Alternative Allele Field](#alternative-allele-field)
* [Genotype Field](#genotype-field)

The following example makes use of [Illumina Platinum Genomes](http://www.illumina.com/platinumgenomes/).  For more detail about how this data was loaded into the Google Genomics API, please see [Google Genomics Public Data](https://cloud.google.com/genomics/data/platinum-genomes).

```{r echo=FALSE, eval=FALSE}
######################[ CHANGE ME ]##################################
# This codelab assumes that the current working directory is where the Rmd file resides.
setwd("/YOUR/PATH/TO/codelabs/R/PlatinumGenomes-QC")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "YOUR-PROJECT-ID"
#####################################################################
```

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
# Set up for BigQuery access.
source("./rHelpers/setup.R")
```

```{r}
# By default this codelab runs upon the Illumina Platinum Genomes Variants.  
# Change the table here if you wish to run these queries against your own data.
queryReplacements <- list("_THE_TABLE_"="genomics-public-data:platinum_genomes.variants")

# To run this against other public data, source in one of the dataset helpers.  For example:
# source("./rHelpers/pgpCGIOnlyDataset.R")
```

## Variants

Let's take a look at a few of the [variants within BRCA1 via BigQuery](https://github.com/googlegenomics/getting-started-bigquery/blob/master/RMarkdown/literate-programming-demo.md#data-visualization):
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("https://raw.githubusercontent.com/googlegenomics/getting-started-bigquery/master/sql/variant-level-data-for-brca1.sql",
                                  project=project,
                                  replacements=queryReplacements)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
DisplayQueryResults(result)
```

These are the variant-level fields common to all variant sets exported to BigQuery from Google Genomics.  There are often dataset-specific variant-level fields as well.  For more information about additional fields, see the schema for the table being queried.  

> In this case, see the Platinum Genomes [variants table schema](https://bigquery.cloud.google.com/table/genomics-public-data:platinum_genomes.variants).

## Non-Variant Segments

Let's take a look at a few non-variant segments within BRCA1:
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/non-variant-segments.sql",
                                  project=project,
                                  replacements=queryReplacements)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
DisplayQueryResults(result)
```

When the data contains non-variant segments, for any analyses that require us to know for example _"how many samples do and do not have a particular SNP?"_, we'll need to make sure that the non-variant segments are considered in addition to the variants.

> The source Platinum Genomes data loaded into the Google Genomics API was in [genome VCF](https://sites.google.com/site/gvcftools/home/about-gvcf/gvcf-conventions) (gVCF) format and therefore has non-variant segments.  

Note that Complete Genomics data also includes non-variant segments and requires the same consideration.

If this query was run on a different dataset and returned no rows, then the data only contains variant records.

## Alternative Allele Field

And then let's take a look at the domain and range of values for alternate_bases:
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/characterize-alts.sql",
                                  project=project,
                                  replacements=queryReplacements)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
DisplayQueryResults(result)
```

> In the case of Platinum Genomes we see from the query results that there are no special charaters in alternate_bases and the maximum length is ~50 base pairs, so just SNPs and small INDELs.

If this query was run on a different dataset, you may wish to run additional queries to understand the domain and range of possible values in the alternate_bases field (e.g., large deletions coded as `<DEL>`, complex structural variants, etc...)

## Genotype Field

And finally let's take a look at the domain and range of values for genotype:
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("./sql/genotypes-brca1.sql",
                                  project=project,
                                  replacements=queryReplacements)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
DisplayQueryResults(result, 10)
```


> In the case of Platinum Genomes we see from the query results the variety of genotypes just within BRCA1:
* no-calls (the -1 values)
* genotypes higher than 1 indicating that the data is not strictly bi-allelic
* genotypes consisting of just a single allele on an autosome

Remove the WHERE clause to run this query over the entire dataset.

# Summary

> To summarize attributes we need to consider when working with Platinum Genomes data:
> * It has non-variant segments which adds complexity above and beyond [similar examples for the 1,000 Genomes dataset](https://github.com/googlegenomics/bigquery-examples/blob/master/1000genomes/sql/README.md).
> * It is comprised only of SNPs and INDELs (contains no structural variants).
> * The values for `alternate_bases` are just comprised of the letters A,C,G,T (e.g., contains no `<DEL>` values).
> * It contains some single-allele and 1/2 genotypes.

--------------------------------------------------------
_Next_: [Part 2: Data Transformation](./Data-Transformation.md)
