<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Comparing Google Genomics Quality Control Results

For details as to how the comparison data was created, see the [provenance details](./README.md).

```{r echo=FALSE, eval=FALSE}
######################[ CHANGE ME ]##################################
# This codelab assumes that the current working directory is where the Rmd file resides.
setwd("/YOUR/PATH/TO/codelabs/R/PlatinumGenomes-QC")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "YOUR-PROJECT-ID"
#####################################################################
```

```{r}
# Set up for BigQuery access.
source("../rHelpers/setup.R")

brca1Replacements <- list("_GENOME_CALL_TABLE_"="genomics-public-data:platinum_genomes.variants",
                     "_MULTISAMPLE_VARIANT_TABLE_"="google.com:biggene:platinum_genomes.expanded_variants",
                     "#_WHERE_"="WHERE reference_name = 'chr17' AND start BETWEEN 41196311 AND 41277499")
```

Sample-Level QC
===============

Sex Inference
-------------
```{r message=FALSE, warning=FALSE, comment=NA}
chrXreplacements <- brca1Replacements
chrXreplacements[["#_WHERE_"]] <- "WHERE
      (reference_name = 'chrX' OR reference_name = 'X')
      # Omit pseudoautosomal regions.
      AND start NOT BETWEEN 59999 AND 2699519
      AND start NOT BETWEEN 154931042 AND 155260559"
result <- DisplayAndDispatchQuery("../sql/homozygous-variants.sql",
                                  project=project,
                                  replacements=chrXreplacements)
```
Number of rows returned by this query: `r nrow(result)`.
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

Comparison with PLINK sex-check results (ignoring pedigree info):
```{r warning=FALSE, results="asis"}
plinkResult <- read.table("./sexCheck/plinkCombined/sex-check-no-fam.sexcheck", header=TRUE)
result <- rename(result, IID=call_call_set_name)
joinedResult <- inner_join(plinkResult, result, by=c("IID"))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(joinedResult), type="html", include.rownames=F)
```

Gender determination based on homozygosity (inbreeding) rate, F, is the same in both methods (considering male for F > 0.5, and female for F < 0).

The slight difference in the F values is due to differences in allele frequency calculations, which results in more SNPs being ignored in PLINK. For instance, ~85,000 of ~345,000 SNPs are excluded in PLINK but only ~41,000 are excluded in the above query (due to having frequency of 1.0 or 0.0). As a concrete example, the SNP in position 2704300 has 15 individuals with 0/0 allele and 2 individuals with 0/1 allele. The minor allele frequency is therefore 2/34 = 0.06, whereas the PLINK allele frequency is 1.0 (actual ratio in PLINK's calculations is 24/24), which results in exclusion of this SNP from the PLINK analysis.

PLINK also considers pedigree info in its analysis. The above results do not use pedigree info, but those results are also provided in [sex-check-with-fam](./sexCheck/plinkCombined/sex-check-with-fam.sexcheck).

Lastly, the multi-sample variants table used in the query only contains reference-matches for SNPs (i.e. indels matching the reference are not included). Thus, indels were also excluded from the above analysis as they would result in inaccurate statistics.

Check Singletons
----------------
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/private-variants-brca1.sql",
                                  project=project,
                                  replacements=brca1Replacements)
```
Number of rows returned by this query: `r nrow(result)`.

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

Compare to [brca1.singletons](./singletons/brca1.singletons) which has 85 some of which are for 0/0 genotypes from reference matching blocks (see the [vcftools command line](./singletons/brca1.log) used to create this file).

```{r}
expectedResult <- read.table("./singletons/brca1.singletons", header=TRUE)
# Convert to zero-based coordinates
expectedResult <- mutate(expectedResult, POS = POS - 1)
# Clean colnames to match
colnames(expectedResult) <- gsub('\\.+', '_', colnames(expectedResult))
```

How many singletons do the two results have in common?
```{r}
nrow(inner_join(result, expectedResult))
```

Which singletons were only identified by BigQuery?
```{r results="asis"}
onlyBQ <- anti_join(result, expectedResult)
print(xtable(onlyBQ), type="html", include.rownames=F)
```

Which singletons were only identified by vcftools?
```{r results="asis"}
onlyVcftools <- anti_join(expectedResult, result)
print(xtable(onlyVcftools), type="html", include.rownames=F)
```

Retrieving the gVCF data for the singletons identified only by vcftools:
```{r message=FALSE, warning=FALSE, comment=NA}
having <- paste("start = ", onlyVcftools$POS,
                sep="", collapse=" OR ")
result <- DisplayAndDispatchQuery("../sql/examine-data.sql",
                                  project=project,
                                  replacements=c(brca1Replacements,
                                                 "_HAVING_"=having))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

It appears that they correspond either to:
* a reference-matching block, so not actually a singleton and just perhaps violating an assumption in the vcftools code
* or a non-singleon variant, perhaps due to a problem in converting the gVCF data to all-positions VCF via gvcftools?

Check Individual Heterozygosity
-----------------------------------

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/homozygous-variants.sql",
                                  project=project,
                                  replacements=brca1Replacements)
```
Number of rows returned by this query: `r nrow(result)`.

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

Compare to [brca1.het](./heterozygous/brca1.het) (see the [vcftools command line](./heterozygous/brca1.log) used to create this file).


```{r}
expectedResult <- read.table("./heterozygous/brca1.het", header=TRUE)
# Clean colnames to match
colnames(expectedResult) <- gsub('\\.+$', '', colnames(expectedResult))
colnames(expectedResult) <- gsub('\\.+', '_', colnames(expectedResult))
```

```{r results="asis"}
result <- dplyr::rename(result, INDV=call_call_set_name)
joinedResult <- inner_join(expectedResult, result, by=c("INDV"))
print(xtable(joinedResult[,order(colnames(joinedResult))]), type="html", include.rownames=F)
```

The logic of the query is the same as vcftools [output_het method](http://sourceforge.net/p/vcftools/code/HEAD/tree/trunk/cpp/variant_file_output.cpp#l165), but the filtering criteria are slightly different, which explains the small difference in the results.

The primary difference is due to using only SNPs in the query whereas vcftools uses both SNPs and indels. This is because the "expanded table" used in the query only contains non-variant segments for SNPs (i.e. indels matching the reference are not included), which results in inaccurate statistics for indels.

There are a few other minor differences as well. For instance, vcftools discards the entire record if a [single sample is non-diploid](http://sourceforge.net/p/vcftools/code/HEAD/tree/trunk/cpp/entry_getters.cpp#l94) whereas the query only discards non-diploid samples within a record. Overall, the filtering differences are minor and the output of the query is largely consistent with vcftools.

Check Identity By State
-----------------------
The Dataflow job was run on variant set 3049512673186936334 over references chr17:41196311:41277499.  See the [Google Genomics Cookbook](http://googlegenomics.readthedocs.org/en/latest/use_cases/compute_identity_by_state/index.html) for the full instructions to run this job.

```{r ibs}
result <- read.table("./identity-by-state/platinum-genomes-brca1-ibs.tsv",
                  col.names=c("sample1", "sample2", "ibsScore", "similar", "observed"))
expectedResult <- read.table("./identity-by-state/brca1-long.ibs",
                  col.names=c("sample1", "sample2", "similar", "observed"))
expectedResult <- mutate(expectedResult, ibsScore= similar / observed)
joinedResult <- inner_join(result, expectedResult, by=c("sample1", "sample2"))
nrow(joinedResult)
ggplot(joinedResult, aes(x=ibsScore.x, y=ibsScore.y)) + geom_point()
model <- lm(ibsScore.y ~ ibsScore.x, joinedResult)
summary(model)
```

There are a few differences between plink pseq IBS and the [Shared Minor Alleles Calculator](https://github.com/googlegenomics/dataflow-java/blob/master/src/main/java/com/google/cloud/genomics/dataflow/functions/SharedMinorAllelesCalculator.java) IBS score calculator in this data Dataflow job.

1. plinkpseq skips variants that are not bi-allelic
1. plinkpseq increments the denominator for calls that are no-calls

Cohort Level QC
===============

Check Hardy-Weinberg Equilibrium
-----------------------------------
```{r message=FALSE, warning=FALSE, comment=NA}
sortAndLimit <- "ORDER BY reference_name, start, alternate_bases"
result <- DisplayAndDispatchQuery("../sql/hardy-weinberg.sql",
                                  project=project,
                                  replacements=c(brca1Replacements,
                                                 "#_ORDER_BY_"=sortAndLimit))
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the first few results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Compare to [brca1.hwe](./hwe/brca1.hwe) (see the [vcftools command line](./hwe/brca1.log) used to create this file).

```{r}
result <- dplyr::rename(result, CHR=reference_name, POS=start)
df <- read.table("./hwe/brca1.hwe", header=TRUE)
obsSplitCol <- "OBS.HOM1.HET.HOM2."
obsTemp <- read.table(text=as.character(df[, obsSplitCol]), sep = "/")
names(obsTemp) <- c("OBS_HOM1", "OBS_HET", "OBS_HOM2")
eSplitCol <- "E.HOM1.HET.HOM2."
eTemp <- read.table(text=as.character(df[, eSplitCol]), sep = "/")
names(eTemp) <- c("E_HOM1", "E_HET", "E_HOM2")
expectedResult <- cbind(cbind(df[setdiff(names(df), c(obsSplitCol,eSplitCol))], obsTemp), eTemp)
# Convert to zero-based coordinates
expectedResult <- mutate(expectedResult, POS = POS - 1)
```

How many results do the two results have in common?
```{r}
nrow(inner_join(result, expectedResult, by=c("CHR", "POS", "OBS_HOM1", "OBS_HET", "OBS_HOM2")))
```

Which results were only identified by BigQuery?
```{r results="asis"}
onlyBQ <- anti_join(result, expectedResult, , by=c("CHR", "POS", "OBS_HOM1", "OBS_HET", "OBS_HOM2"))
print(xtable(arrange(onlyBQ, CHR, POS)), type="html", include.rownames=F)
```

Note vcftools appears to skip variants with single allele genotypes:
```
zgrep 41242078 platinum_genomes_brca1_expanded_merged.vcf.gz 
chr17  41242078  .  G	A	143	LowGQX;TruthSensitivityTranche99.90to100.00;LowQD;SiteConflict	BLOCKAVG_min30p3a;MQ=57;MQ0=0;BaseQRankSum=0.781;Dels=0.3;FS=1.561;HRun=11;HaplotypeScore=77.7361;MQRankSum=0.093;QD=2.01;ReadPosRankSum=-2.871;SB=-45.67;VQSLOD=-1.8762;culprit=QD;set=FilteredInAll;DP=425;AF=0.5;AN=25;AC=1	GT:DP:GQX:MQ:AD:GQ:PL:VF	0/0:57:99:59:.:.:.:.	0:27:25:57:26:25.38:.:.	0/0:51:99:57:.:.:.:.	0/1:50:99:59:42,8:99:173,0,1238:0.16	0/0:46:99:59:.:.:.:.	0/0:44:99:60:.:.:.:.	.:46:.:59:40,6:.:.:.	0/0:40:85:59:.:.:.:.	0/0:40:85:59:.:.:.:.	0/0:63:99:58:.:.:.:.	0:42:2:58:37:1.58:.:.	0:33:0:57:29:0.03:.:.	.:44:.:58:31,12:.:.:.	0/0:44:90:58:.:.:.:.	0/0:40:87:58:.:.:.:.	.:44:.:57:39,5:.:.:.	0/0:55:99:59:.:.:.:.
```

Which results were only identified by vcftools?
```{r results="asis"}
onlyVcftools <- anti_join(expectedResult, result, , by=c("CHR", "POS", "OBS_HOM1", "OBS_HET", "OBS_HOM2"))
print(xtable(arrange(onlyVcftools, CHR, POS)), type="html", include.rownames=F)
```

Retrieving the gVCF data for the results identified only by vcftools:
```{r message=FALSE, warning=FALSE, comment=NA}
having <- paste("start <= ", onlyVcftools$POS,
                "AND",
                "end >= ", onlyVcftools$POS+1)
result <- DisplayAndDispatchQuery("../sql/examine-data.sql",
                                  project=project,
                                  replacements=c(brca1Replacements,
                                                 "_HAVING_"=having))
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

It appears that with BigQuery we are computing HWE for all the same variants as vcftools and the expected and Chi-Squared values are only slightly different.

See also: the [gVCF version of this query](../sql/hardy-weinberg-brca1.sql), which is close but only works for SNPs and needs a RIGHT OUTER JOIN to compute values for variants for which all the samples have the variant.


===============

Check Transition-Transversion Ratio
-----------------------------------
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/ti-tv-ratio.sql",
                                  project=project,
                                  replacements=c(brca1Replacements,
                                                 "_WINDOW_SIZE_"="1000000"))
```
The result:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

Let's compare this to what we get from vcftools.  For information about the vcftools command see the [log](./titv/platinum_genomes_brca1_expanded_merged.log).  
```{r}
expectedResult <- read.table("./titv/platinum_genomes_brca1_expanded_merged.TsTv.summary", header=TRUE)
```
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(expectedResult), type="html", include.rownames=F)
```
We can see that with BigQuery we get 143 transition mutations, and 132 transversion mutations.  Using vcftools we get two less of each category, 141 transitions and 130 transversions.  


Let's figure out what the differences are.  First, we need to get the specific mutations from BigQuery.
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/ti-tv-variants.sql",
                                  project=project,
                                  replacements=brca1Replacements)
```
Here's the first few variants reported by BigQuery:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```
Next, import the [transition and transversion mutations](./titv/brca.titv.csv) from the raw vcf file.  These were pulled out of the original vcf using a [custom perl script](./titv/pull_titv.pl).  vcftools does not output a new vcf file with only the transitions and transversions so we need to use a proxy method.
```{r}
expectedResult <- read.csv("./titv/brca.titv.csv", header=FALSE)
# Set column names
names(expectedResult) <- c("reference_name","start","reference_bases","alternate_bases")
# Convert to zero-based coordinates
expectedResult <- mutate(expectedResult, start = start - 1)
```
Which variants were only identified by BigQuery?
```{r}
onlyBQ <- anti_join(result, expectedResult)
```

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(onlyBQ), type="html", include.rownames=F)
```
The four variants identified only by BigQuery are from two positions, each having multiple alternate alleles.  The perl script used to identify positions with transitions and transversion did not account for this, it is likely vcftools does not either.  Because vcftools does not output we cannot say for sure whether these are the 4 variants that vcftools missed, but it is a safe assumption given that we have two additional transitions and two transversions from these positions (which matches the discrepencey we originally had) as well as a logical reason for a bug in vcftools.

Let's double check that no variants were identified only by vcftools.
```{r}
nrow(anti_join(expectedResult, result))
```

